File: ios/metro_drone_plugin/Sources/metro_drone_plugin/Handlers/MetronomeStreamHandler.swift
import Flutter
import UIKit

public class MetronomeStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private let metronome: Metronome

    init (metronome: Metronome) {
        self.metronome = metronome
        super.init()
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        self.metronome.onFieldUpdated = { [weak self] (field: String, value: Any) in
            self?.eventSink?([field: value])
        }

        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.metronome.onFieldUpdated = nil
        self.eventSink = nil
        return nil
    }
}

public class MetronomeTickStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private let metronome: Metronome

    init (metronome: Metronome) {
        self.metronome = metronome
        super.init()
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        self.metronome.onTickUpdated = { [weak self] (tickIndex: Int) in
            self?.eventSink?(tickIndex)
        }

        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.metronome.onTickUpdated = nil
        self.eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/io/modacity/metro_drone_plugin/handlers/MetronomeStreamHandler.kt
package io.modacity.metro_drone_plugin.handlers

import app.metrodrone.domain.metronome.Metronome
import io.flutter.plugin.common.EventChannel
import android.util.Log

class MetronomeStreamHandler(private var metronome: Metronome) : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        metronome.onFieldUpdate = { field, value ->
            sendUpdate(field, value)
        }
    }

    override fun onCancel(arguments: Any?) {
        metronome.onFieldUpdate = null
        eventSink = null
    }

    // Helper method to send updates to Flutter
    fun sendUpdate(field: String, value: Any) {
        Log.d("Metronome", "${field}: $value")
        eventSink?.success(mapOf(field to value))
    }

    // Helper method to send multiple field updates
    fun sendUpdates(updates: Map<String, Any>) {
        eventSink?.success(updates)
    }
}

class MetronomeTickStreamHandler(private var metronome: Metronome) : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        metronome.onTickUpdated = { tick -> sendTick(tick) }
    }

    override fun onCancel(arguments: Any?) {
        metronome.onTickUpdated = null
        eventSink = null
    }

    // Helper method to send tick updates to Flutter
    fun sendTick(tickIndex: Int) {
        eventSink?.success(tickIndex)
    }
}